spec/API_specification/signatures/array_object.py

```diff
@@ -0,0 +1,942 @@
+
+        - The returned value should be a tuple; however, where warranted, an array library may choose to return a custom shape object. If an array library returns a custom shape object, the object must be immutable, must support indexing for dimension retrieval, and must behave similarly to a tuple.
+        """
+
+        self.size = None
+        """
+        Number of elements in an array.
+
+        Returns
+        -------
+        out: Optional[int]
+            number of elements in an array. The returned value must be ``None`` if and only if one or more array dimensions are unknown.
+
+        Notes
+        -----
+        - This must equal the product of the array's dimensions.
+
+        - For array libraries having graph-based computational models, an array may have unknown dimensions due to data-dependent operations.
+        """
+
+        self.T = None
+        """
+        Transpose of the array.
+
+        The array instance must be two-dimensional. If the array instance is not two-dimensional, an error should be raised.
+
+        Returns
+        -------
+        out: array
+            two-dimensional array whose first and last dimensions (axes) are permuted in reverse order relative to original array. The returned array must have the same data type as the original array.two-dimensional array whose first and last dimensions (axes) are permuted in reverse order relative to original array. The returned array must have the same data type as the original array.
+
+        Notes
+        -----
+        - Limiting the transpose to two-dimensional arrays (matrices) deviates from the NumPy et al practice of reversing all axes for arrays having more than two-dimensions. This is intentional, as reversing all axes was found to be problematic (e.g., conflicting with the mathematical definition of a transpose which is limited to matrices; not operating on batches of matrices; et cetera). In order to reverse all axes, one is recommended to use the functional ``permute_dims`` interface found in this specification.
+        """
+
+    def __abs__(self: array, /) -> array:
+        """
+        Calculates the absolute value for each element of an array instance (i.e., the element-wise result has the same magnitude as the respective element but has positive sign).
+
+        Parameters
+        ----------
+        self: array
+            array instance. Should have a numeric data type.
+
+        Returns
+        -------
+        out: array
+            an array containing the element-wise absolute value. The returned array must have the same data type as ``self``.
+
+        Notes
+        -----
+        - For signed integer data types, the absolute value of the minimum representable integer is implementation-dependent.
+
+        - Element-wise results must equal the results returned by the equivalent element-wise function :ref:`function-abs`.
```

Comment:
The cross-references here need to be fixed, e.g., this should be ``` `:func:`abs` ```. The build shows warnings for all of these. We should consider building with `-W` so that the warnings become errors.

Reply:
The issue with changing all these references right now, is that they are mapping to the `element-wise` category, which isn't in rst yet. So they will look like this,

![image](https://user-images.githubusercontent.com/20992645/148790579-5426b93f-9eeb-4492-a42c-315030ec0455.png)

When I was working on this, I decided to leave them like that because the link will work, and then I will change all these references in the `element-wise` PR. In this way, the specification page will not have broken links.

Reply:
Ah right. When I reviewed these, I just merged all your branches together so that I could look at everything at once.